src/main/java/io/cdap/plugin/gcp/bigquery/source/PartitionedBigQueryInputFormat.java

```diff
@@ -0,0 +1,192 @@
+                                                                                serviceFilePath);
+    StandardTableDefinition tableDefinition = Objects.requireNonNull(sourceTable).getDefinition();
+    TimePartitioning timePartitioning = tableDefinition.getTimePartitioning();
+    if (timePartitioning == null) {
+      return null;
+    }
+    StringBuilder condition = new StringBuilder();
+    String columnName = timePartitioning.getField() != null ? timePartitioning.getField() : DEFAULT_COLUMN_NAME;
+
+    LegacySQLTypeName columnType = null;
+    if (!DEFAULT_COLUMN_NAME.equals(columnName)) {
+      columnType = tableDefinition.getSchema().getFields().get(columnName).getType();
+    }
+
+    if (partitionFromDate != null) {
+      if (LegacySQLTypeName.DATE.equals(columnType)) {
+        condition.append("TIMESTAMP(").append(columnName).append(")");
+      } else {
+        condition.append(columnName);
+      }
+      condition.append(" >= ").append("TIMESTAMP(\"").append(partitionFromDate).append("\")");
+    }
+    if (partitionFromDate != null && partitionToDate != null) {
+      condition.append(" and ");
+    }
+    if (partitionToDate != null) {
+      if (LegacySQLTypeName.DATE.equals(columnType)) {
+        condition.append("TIMESTAMP(").append(columnName).append(")");
+      } else {
+        condition.append(columnName);
+      }
+      condition.append(" <= ").append("TIMESTAMP(\"").append(partitionToDate).append("\")");
+    }
+
+    String tableName = dataset + "." + table;
+    return String.format(queryTemplate, tableName, condition.toString());
+  }
+
+  private static void runQuery(
+    BigQueryHelper bigQueryHelper, String projectId, TableReference tableRef, String query)
+    throws IOException, InterruptedException {
+
+    // Create a query statement and query request object.
+    JobConfigurationQuery queryConfig = new JobConfigurationQuery();
+    queryConfig.setAllowLargeResults(true);
+    queryConfig.setQuery(query);
+    queryConfig.setUseLegacySql(false);
+
+    // Set the table to put results into.
+    queryConfig.setDestinationTable(tableRef);
+
+    queryConfig.setCreateDisposition("CREATE_IF_NEEDED");
+
+    // Require table to be empty.
+    queryConfig.setWriteDisposition("WRITE_EMPTY");
+
+    JobConfiguration config = new JobConfiguration();
+    config.setQuery(queryConfig);
+
+    JobReference jobReference =
+      bigQueryHelper.createJobReference(projectId, "querybasedexport", null);
+
+    Job job = new Job();
+    job.setConfiguration(config);
+    job.setJobReference(jobReference);
+
+    // Run the job.
+    Job response = bigQueryHelper.insertJobOrFetchDuplicate(projectId, job);
+
+    // Create anonymous Progressable object
+    Progressable progressable = new Progressable() {
+      @Override
+      public void progress() {
+        // TODO(user): ensure task doesn't time out
+      }
+    };
+
+    // Poll until job is complete.
+    BigQueryUtils.waitForJobCompletion(
```

Comment:
we've had bugs in the past with datasets in non-default regions. Let's be sure to test with datasets in different regions.

Reply:
I will check this case.

Reply:
@albertshau I checked with non-default regions and you was right I got error:
com.google.api.client.googleapis.json.GoogleJsonResponseException: 404 Not Found
{
  "code" : 404,
  "errors" : [ {
    "domain" : "global",
    "message" : "Not found: Job cdap-cybervision:querybasedexport-eff333bb-3784-4e1d-ac2b-ac67683c24bd",
    "reason" : "notFound"
  } ],
  "message" : "Not found: Job cdap-cybervision:querybasedexport-eff333bb-3784-4e1d-ac2b-ac67683c24bd",
  "status" : "NOT_FOUND"
}

If bugs with non-standard regions are known, are there any solutions that can be used?

Reply:
I've merged release to develop, can you check if the 404 problem still exists?

For the problem with the GCS bucket being created in the wrong region, was that only a problem in preview? I think it's because of:

```
if (!context.isPreviewEnabled()) {
      BigQueryUtil.createResources(bigQuery, GCPUtils.getStorage(config.getDatasetProject(), credentials),
                                   config.getDataset(), bucket);
    }
```

The if condition should be removed, since the GCS bucket is required whether it's in preview or not.